src/hpa.c

```diff
@@ -0,0 +1,395 @@
+	malloc_mutex_lock(tsdn, &shard->mtx);
+	psset_alloc_new(&shard->psset, grow_edata, edata, size);
+	malloc_mutex_unlock(tsdn, &shard->mtx);
+
+	malloc_mutex_unlock(tsdn, &shard->grow_mtx);
+	return edata;
+}
+
+static edata_t *
+hpa_alloc(tsdn_t *tsdn, pai_t *self, size_t size,
+    size_t alignment, bool zero) {
+
+	assert((size & PAGE_MASK) == 0);
+	/* We don't handle alignment or zeroing for now. */
+	if (alignment > PAGE) {
+		return NULL;
+	}
+	if (zero) {
+		return NULL;
+	}
+
+	witness_assert_depth_to_rank(tsdn_witness_tsdp_get(tsdn),
+	    WITNESS_RANK_CORE, 0);
+
+	hpa_shard_t *shard = (hpa_shard_t *)self;
+
+	edata_t *edata;
+	if (size <= shard->ps_alloc_max) {
+		edata = hpa_alloc_psset(tsdn, shard, size);
+		emap_register_boundary(tsdn, shard->hpa->central.emap, edata,
+		    SC_NSIZES, /* slab */ false);
+	} else {
+		edata = hpa_alloc_central(tsdn, shard, size, size);
+	}
+
+	witness_assert_depth_to_rank(tsdn_witness_tsdp_get(tsdn),
+	    WITNESS_RANK_CORE, 0);
+	if (edata != NULL) {
+		emap_assert_mapped(tsdn, shard->hpa->central.emap, edata);
+		assert(edata_pai_get(edata) == EXTENT_PAI_HPA);
+		assert(edata_state_get(edata) == extent_state_active);
+		assert(edata_arena_ind_get(edata) == shard->ind);
+		assert(edata_szind_get_maybe_invalid(edata) == SC_NSIZES);
+		assert(!edata_slab_get(edata));
+		assert(edata_committed_get(edata));
+		assert(edata_base_get(edata) == edata_addr_get(edata));
+		assert(edata_base_get(edata) != NULL);
+	}
+	return edata;
+}
+
+static bool
+hpa_expand(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    size_t old_size, size_t new_size, bool zero) {
+	/* Expand not yet supported. */
+	return true;
+}
+
+static bool
+hpa_shrink(tsdn_t *tsdn, pai_t *self, edata_t *edata,
+    size_t old_size, size_t new_size) {
+	/* Shrink not yet supported. */
+	return true;
+}
+
+static void
+hpa_dalloc_central(tsdn_t *tsdn, hpa_shard_t *shard, edata_t *edata) {
+	hpa_t *hpa = shard->hpa;
+
+	edata_arena_ind_set(edata, hpa->ind);
+	malloc_mutex_lock(tsdn, &hpa->mtx);
+	hpa_central_dalloc(tsdn, &hpa->central, edata);
+	malloc_mutex_unlock(tsdn, &hpa->mtx);
+}
+
+static void
+hpa_dalloc(tsdn_t *tsdn, pai_t *self, edata_t *edata) {
+	hpa_shard_t *shard = (hpa_shard_t *)self;
+
+	edata_addr_set(edata, edata_base_get(edata));
+	edata_zeroed_set(edata, false);
+
+	assert(edata_pai_get(edata) == EXTENT_PAI_HPA);
+	assert(edata_state_get(edata) == extent_state_active);
+	assert(edata_arena_ind_get(edata) == shard->ind);
+	assert(edata_szind_get_maybe_invalid(edata) == SC_NSIZES);
+	assert(!edata_slab_get(edata));
+	assert(edata_committed_get(edata));
+	assert(edata_base_get(edata) != NULL);
+
+	if (edata_ps_get(edata) != NULL) {
+		emap_deregister_boundary(tsdn, shard->hpa->central.emap, edata);
+
+		malloc_mutex_lock(tsdn, &shard->mtx);
+		edata_t *evicted_ps = psset_dalloc(&shard->psset, edata);
+		malloc_mutex_unlock(tsdn, &shard->mtx);
+
+		edata_cache_put(tsdn, shard->edata_cache, edata);
+
+		if (evicted_ps != NULL) {
```

Comment:
The code logic here is pretty smooth and easy to follow. I'd still say to add some high level comments somewhere about the dalloc logic, e.g. what / when goes to shard vs hpa_central.

Reply:
Added some comments.